### PR TITLE
Fix #596: Allow file/folder download

### LINKED ISSUE #596: Allow users to download individual file or folder structure without having to download the whole project.
Feature first filed here mozilla/thimble.mozilla.org#1313 by @ScottDowne 

It would be amazing if brackets had an option to download individual file or folder structure without having to download the whole project.

Here are the behaviors for this new feature:

## Downloading a Single file
**UI**:
download_file
**Expected behavior**: The selected file will be downloaded without it being zipped.

## Downloading a Folder
**UI:**
download_folder
**Expected behavior:** This will zip the folder structure/contents and download it. The contents in the zip file will look like this: `{ folder name } / { file contents }`

If the folder is empty and the user chooses to download it, should an error message show or should the empty folder be zipped and downloaded? What do you think?

### PR DESCRIPTION
**Pull request by simon66**
Fixed issue mozilla#596

Here's a demo of a file download then a folder download:

ezgif com-video-to-gif

I made it so that my modifications doesn't affect Thimble (Since it uses the `archive` function to download the whole project).

PS: Can I add CMD_FILE_DOWNLOAD to the other locales file (Thus the title [WIP] haha)? Not too sure what are your requirements when working with locales :)

### CODE REVIEW

File: src/nls/root/strings.js [hunk 1]
```diff
@@ -331,6 +331,7 @@ define({
     "CMD_PROJECT_SETTINGS"                : "Project Settings\u2026",
     "CMD_FILE_RENAME"                     : "Rename",
     "CMD_FILE_DELETE"                     : "Delete",
+    "CMD_FILE_DOWNLOAD"                   : "Download",
```
Comment (simon66) [inline src/nls/root/strings.js]: I might be wrong adding this String here. I never worked with localles before.

**Comment by humphd** [comment]
Let me know when you want this reviewed.  One thing I know I'm going to tell you is that you need to rebase your code and update your master branch, because it looks like you're behind by quite a bit.

**Comment by simon66** [comment]
Thanks @humphd 
I made the changes and pushed them (Code is ready for review 👍 ).
As for locales, I've only added `CMD_FILE_DOWNLOAD` to `en-US`. Don't I need to add `CMD_FILE_DOWNLOAD` to the other files? Or is that something that needs to be added in the future?

Thanks!

File: src/config.json [hunk 2]
```diff
@@ -102,4 +102,4 @@
         "start": "http-server -p 8000 --cors"
```
Comment (humphd) [inline src/config.json]: Revert this with `git checkout master src/config.json`

File: src/document/DocumentCommandHandlers.js [hunk 3]
```diff
@@ -1596,6 +1597,12 @@ define(function (require, exports, module) {
         ProjectManager.showInTree(MainViewManager.getCurrentlyViewedFile(MainViewManager.ACTIVE_PANE));
     }
 
+    /** Download selected file or folder structure **/
+    function handleFileDownload() {
+        var entry = ProjectManager.getSelectedItem();
+        ArchiveUtils.archiveWithPath(entry._path,function(){});
```
Comment (humphd) [inline src/document/DocumentCommandHandlers.js]: Depending on what this is (e.g., file vs. folder) I think we should do different things.  I think it's odd to click "Download" on a file and have it send me a `.zip`.  What I'd do is check to see if this is a file vs. folder, and based on that do one of two things:

* For the file case: generalize my code here https://github.com/simon66/brackets/blob/70a2290e4e64ce5e49fe3a3398c11c852d2939f2/src/filesystem/impls/filer/ArchiveUtils.js#L208-L210 to have it get the blob URL from `BlobUtils` for this path and then use that with `saveAs()`.  Then you'll just get the actual file you want.
* If it's a dir, I think archiving is probably the right move, but we should probably alter what we send.  Fixing the "project/" bug here might make sense, since we should just send the folder zipped.

I also don't love how you're passing an empty callback.  It would be better to check for `callback` in `archiveWithPath` and if it's not there, create it then.

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 4]
```diff
@@ -263,6 +269,7 @@ define(function (require, exports, module) {
     }
 
     exports.archive = archive;
+    exports.archiveWithPath = archiveWithPath;
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: Let's just change how `archive` works, so it always takes a path, and fix every call site that is currently not passing one in.

File: locales/en-US/editor.properties [hunk 5]
```diff
@@ -116,6 +116,7 @@ CMD_FILE_NEW=New File
 CMD_FILE_NEW_FOLDER=New Folder
 CMD_FILE_RENAME=Rename
 CMD_FILE_DELETE=Delete
+CMD_FILE_DOWNLOAD=Download
```
Comment (humphd) [inline locales/en-US/editor.properties]: Talk to @gideonthomas about the dance we need to do for this string, since it also needs to get added to Thimble.

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 6]
```diff
@@ -141,10 +141,16 @@ define(function (require, exports, module) {
             _unzip(zipfile);
         }
     }
-
+
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: Revert this change.

File: src/document/DocumentCommandHandlers.js (hunk 3 above)
Comment (simon66) [inline src/document/DocumentCommandHandlers.js]: Thanks for the feedback.
As I'm applying these changes, I noticed that `saveAs()` doesn't take a blobURL as a parameter.

screen shot 2017-04-06 at 10 52 32 am

screen shot 2017-04-06 at 10 52 53 am

Is there a step that I'm missing?

Comment (humphd) [inline src/document/DocumentCommandHandlers.js]: That sucks.  Well, since it takes a `Blob` vs. a BlobURL (i.e., it wants to create the URL for the Blob itself), you'll just have to do that instead (e.g., read the file's contents, get its mime type, and create a new `Blob` yourself).

**Comment by simon66** [comment]
I hate to fix the MIME-TYPE and deal with the locales 👍

File: src/extensions/default/bramble/lib/RemoteCommandHandler.js [hunk 7]
```diff
@@ -150,7 +151,7 @@ define(function (require, exports, module) {
             break;
         case "BRAMBLE_EXPORT":
             skipCallback = true;
-            ArchiveUtils.archive(callback);
+            ArchiveUtils.archive(StartupState.project("root"),callback);
```
Comment (humphd) [inline src/extensions/default/bramble/lib/RemoteCommandHandler.js]: Space after comma please.

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 8]
```diff
@@ -142,11 +143,13 @@ define(function (require, exports, module) {
...
     }
 
-    // Zip the entire project, starting at the project root.
-    function archive(callback) {
+    // Zip specific file or folder structure
+    function archive(path, callback) {
         var root = StartupState.project("root");
         var rootRegex = new RegExp("^" + root + "\/?");
-
+        if (callback === undefined) {
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: `callback = callback || function(){};`

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 9]
```diff
@@ -194,15 +197,32 @@ define(function (require, exports, module) {
             });
         }
 
-        add(root, function(err) {
-            if(err) {
-                return callback(err);
+        // Check if path is a directory or not
+        fs.stat(path, function(err, stats) {
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: Needs error handling

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 10]
```diff
@@ -194,15 +197,32 @@ define(function (require, exports, module) {
...
+                    }
+                    // Prepare folder for download
+                    var compressed = jszip.generate({type: 'arraybuffer'});
+                    var blob = new Blob([compressed], {type: "application/zip"});
+                    saveAs(blob, "project.zip");
+                    callback();
+                });
+            } else {
+                // Prepare file for download
+                fs.readFile(path, {encoding: null}, function(err, data){
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: I don't think it makes sense to do this inside a function called `archive`.  If I pass you a `path` to `archive`, you should always produce a `.zip`.  Instead of this, you should check at your call site to see if this is a file or folder, and only call `archive` in the latter case, otherwise do this code.

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 11]
```diff
@@ -194,15 +197,32 @@ define(function (require, exports, module) {
...
+                    if (err) {
+                       return callback(err);
+                    }
+                    
+                    var blob = new Blob([data], {type: "text/plain"});
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: use https://github.com/mozilla/brackets/blob/master/src/filesystem/impls/filer/lib/content.js#L40 for the mime-type.

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 12]
```diff
@@ -194,15 +197,32 @@ define(function (require, exports, module) {
...
+                    var filename = path.substring(path.lastIndexOf("/") + 1, path.length);
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: use `Path.basename()`

File: src/document/DocumentCommandHandlers.js [hunk 13]
```diff
@@ -1596,6 +1600,34 @@ define(function (require, exports, module) {
         ProjectManager.showInTree(MainViewManager.getCurrentlyViewedFile(MainViewManager.ACTIVE_PANE));
     }
 
+    /** Download selected file or folder structure **/
+    function handleFileDownload() {
+        var selectedItem = ProjectManager.getSelectedItem();
+        var path = selectedItem._path;
+        try {
```
Comment (humphd) [inline src/document/DocumentCommandHandlers.js]: Why would anything in here throw?

File: src/document/DocumentCommandHandlers.js [hunk 14]
```diff
@@ -1596,6 +1600,34 @@ define(function (require, exports, module) {
...
+            fs.stat(path, function(err, stats) {
+                if (stats.type === "DIRECTORY") {
```
Comment (humphd) [inline src/document/DocumentCommandHandlers.js]: You need to check `if(err) {...}` here, and probably log something or show a modal dialog indicating that there was a problem.

File: src/document/DocumentCommandHandlers.js [hunk 15]
```diff
@@ -1596,6 +1600,34 @@ define(function (require, exports, module) {
...
+                    ArchiveUtils.archive(path);
```
Comment (humphd) [inline src/document/DocumentCommandHandlers.js]: do `return ArchiveUtils.achive(path);` here so you don't need the `else` below, or the indenting.

File: src/document/DocumentCommandHandlers.js [hunk 16]
```diff
@@ -1596,6 +1600,34 @@ define(function (require, exports, module) {
...
+                } else {
+                    // Prepare file for download
+                    fs.readFile(path, {encoding: null}, function(err, data){
+                        if (err) {
+                            console.error(err);
```
Comment (humphd) [inline src/document/DocumentCommandHandlers.js]: Same as above, you probably want to show a modal dialog in brackets that says there was a problem.

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 17]
```diff
@@ -193,12 +193,12 @@ define(function (require, exports, module) {
                 }
             });
         }
-
-        add(root, function(err) {
+
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: Get rid of this.

File: locales/en-US/editor.properties (hunk 5 above)
Comment (gideonthomas) [inline locales/en-US/editor.properties]: All you need to do is also add the same line to the `locales/en-US/editor.properties` file in the thimble repository and open up a PR for it. If you look closely, that file is identical to this one.

Comment (simon66) [inline locales/en-US/editor.properties]: Ok, I'll do a PR on thimble with the fix mentioned.
Also, I added 2 more strings
`UNEXPECTED_ERROR` and `ERROR_DIALOG_HEADER`

**Comment by simon66** [comment]
hi @humphd,

I made the changed as per your request 👍

**Comment by flukeout** [comment]
Testing

**Comment by flukeout** [comment]
Ok, this works great for both Chrome & Firefox when I run it from the ``/dist`` folder on localhost like - ``http://localhost:8000/dist/hosted.html`` but not in ``/src`` - is that expected behaviour?

**Comment by humphd** [comment]
What does it do in `src/`?  Is it just missing strings?  Did you run `npm run localize`?

**Comment by flukeout** [comment]
``npm run localize`` made it work in ``src`` also - didn't realize that was the missing step!

**Comment by humphd** [comment]
We have new code on master that does this automatically when you run the server, but I think this branch is behind and doesn't have it.

**Comment by simon66** [comment]
@humphd is correct. This branch doesn't have the new changes. Should I update it?

**Comment by humphd** [comment]
You can rebase, sure.

**Comment by simon66** [comment]
Hi @humphd, I just did a `rebase` on my branch.

File: locales/en-US/editor.properties [hunk 18]
```diff
@@ -4,8 +4,12 @@
 
 # General File I/O Error strings
 
+# Dialog Headers
+ERROR_DIALOG_HEADER=Error
```
Comment (humphd) [inline locales/en-US/editor.properties]: Is this really necessary?  Isn't there already a string we could use, or doesn't Brackets have some way to title an error dialog?

File: src/document/DocumentCommandHandlers.js [hunk 19]
```diff
@@ -1598,6 +1602,55 @@ define(function (require, exports, module) {
         ProjectManager.showInTree(MainViewManager.getCurrentlyViewedFile(MainViewManager.ACTIVE_PANE));
     }
 
+    /** 
+     * Show error dialog
+     * @param {string} err Error message to be displayed to the user
+     * @param {callback} callback
+     */
+    function showErrorDialog(err, callback){
+        callback = callback || function () {};
```
Comment (humphd) [inline src/document/DocumentCommandHandlers.js]: Do we really need to support the case of having no callback?  Aren't you the only caller of this?

File: src/document/DocumentCommandHandlers.js [hunk 20]
```diff
@@ -1598,6 +1602,55 @@ define(function (require, exports, module) {
...
+                Strings.UNEXPECTED_ERROR,
+                err),
+            [{
+                className : Dialogs.DIALOG_BTN_CLASS_PRIMARY,
+                id        : Dialogs.DIALOG_BTN_OK,
+                text      : Strings.OK
+            }]
+        );
+        console.error(err);
+        callback();
```
Comment (humphd) [inline src/document/DocumentCommandHandlers.js]: You don't want to wait until the dialog is dismissed?  If so, you can get a promise from the dialog and wait on that to finish.

File: src/document/DocumentCommandHandlers.js [hunk 21]
```diff
@@ -1598,6 +1602,55 @@ define(function (require, exports, module) {
...
+            StringUtils.format(
+                Strings.UNEXPECTED_ERROR,
+                err),
+            [{
+                className : Dialogs.DIALOG_BTN_CLASS_PRIMARY,
+                id        : Dialogs.DIALOG_BTN_OK,
+                text      : Strings.OK
+            }]
+        );
+        console.error(err);
```
Comment (humphd) [inline src/document/DocumentCommandHandlers.js]: I don't think we need this.

File: src/document/DocumentCommandHandlers.js [hunk 22]
```diff
@@ -1598,6 +1602,55 @@ define(function (require, exports, module) {
...
+    /** Download selected file or folder structure **/
+    function handleFileDownload() {
+        var selectedItem = ProjectManager.getSelectedItem();
+        var path = selectedItem._path;
+        fs.stat(path, function(err, stats) {
+            if (err) {
+                showErrorDialog(err);
+                return;
+            }
+            if (stats.type === "DIRECTORY") {
```
Comment (humphd) [inline src/document/DocumentCommandHandlers.js]: Empty line above this `if` to set it off from the previous section.

File: src/document/DocumentCommandHandlers.js [hunk 23]
```diff
@@ -1598,6 +1602,55 @@ define(function (require, exports, module) {
...
+                return ArchiveUtils.archive(path);
+            }
+            // Prepare file for download
```
Comment (humphd) [inline src/document/DocumentCommandHandlers.js]: Empty line above this, same reason as above.

File: src/extensions/default/bramble/lib/RemoteCommandHandler.js [hunk 24]
```diff
@@ -227,4 +228,4 @@ define(function (require, exports, module) {
     }
 
     exports.handleRequest = handleRequest;
-});
+});
```
Comment (humphd) [inline src/extensions/default/bramble/lib/RemoteCommandHandler.js]: Undo this change please, always include a newline at the end of files.

**Comment by gideonthomas** [comment]
@simon66 any updates on this?

**Comment by simon66** [comment]
Oh sorry, I didn't see the notification for this item. Going to make the changes and push them.

**Comment by humphd** [comment]
@simon66 how is this coming?  Can we finish it this week please?

File: locales/en-US/editor.properties (hunk 18 above)
Comment (simon66) [inline locales/en-US/editor.properties]: I couldn't find any string with the value "Error" or similar. The closest thing I found was "Genetic_error", but that string translate to "(error )" (which looks weird).

**Comment by simon66** [comment]
Hi @humphd, just pushed the new changes.

**Comment by humphd** [comment]
OK, great.  We've just frozen the code so we can launch next week.  After the freeze, I'll come back to this.

**Comment by humphd** [comment]
I'm finishing this in https://github.com/mozilla/brackets/pull/780
